application/controllers/FailedEmailController.php

```diff
@@ -0,0 +1,179 @@
+                        "data" => [
+                            'success' => true,
+                            'html'    => $this->renderPartial('partials/modal/resend_result', [
+                                'successfullEmailCount' => $result['successfullEmailCount'],
+                                'failedEmailCount'      => $result['failedEmailCount']
+                            ], true)
+                        ]
+                    ]);
+                }
+            }
+            return $this->renderPartial('/admin/super/_renderJson', [
+                "data" => [
+                    'success' => false,
+                    'message' => gT('No match could be found for selection'),
+                ]
+            ]);
+        }
+        return $this->renderPartial('/admin/super/_renderJson', [
+            "data" => [
+                'success' => false,
+                'message' => gT('Please select at least one item'),
+            ]
+        ]);
+    }
+
+    /**
+     * @throws CHttpException|CException
+     */
+    public function actionDelete(): string
+    {
+        $surveyId = sanitize_int(App()->request->getParam('surveyid'));
+        if (!$surveyId) {
+            throw new CHttpException(403, gT("Invalid survey ID"));
+        }
+        if (!Permission::model()->hasSurveyPermission($surveyId, 'responses', 'delete')) {
+            App()->user->setFlash('error', gT("You do not have permission to access this page."));
+            $this->redirect(['failedemail/index/', 'surveyid' => $surveyId]);
+        }
+        $item = [App()->request->getParam('item')];
+        $items = json_decode(App()->request->getParam('sItems'));
+        $selectedItems = $items ?? $item;
+        if (!empty($selectedItems)) {
+            $criteria = new CDbCriteria();
+            $criteria->select = 'id, email_type, recipient';
+            $criteria->addCondition('surveyid', $surveyId);
+            $criteria->addInCondition('id', $selectedItems);
+            $failedEmails = new FailedEmail();
+            $deletedCount = $failedEmails->deleteAll($criteria);
+            if ($items) {
+                return $this->renderPartial('partials/modal/delete_result_body', [
+                    'deletedCount' => $deletedCount,
+                ]);
+            }
+            return $this->renderPartial('/admin/super/_renderJson', [
+                "data" => [
+                    'success' => true,
+                    'html'    => $this->renderPartial('partials/modal/delete_result', [
+                        'deletedCount' => $deletedCount,
+                    ], true),
+                ]
+            ]);
+        }
+        return $this->renderPartial('/admin/super/_renderJson', [
+            "data" => [
+                'success' => false,
+                'message' => gT('Please select at least one item'),
+            ]
+        ]);
+    }
+
+    public function actionModalContent(): string
```

Comment:
Missing permission check?

Reply:
you want an additional check for the content? there is allready one for displaying the button

Reply:
Yep, since a malicious user can construct a custom http request to fetch data.

Reply:
> you want an additional check for the content? there is allready one for displaying the button

All controller public function need permission check. Else : security issue.
With yii : all controller public function can be call by direct url